tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/MicroSecEndTest.java

```diff
@@ -0,0 +1,296 @@
+        if (hasDataCharecteristicViolation(node, "Stereotype", "entrypoint", "Stereotype", "encrypted_connection")) {
+            violationsSet.add(7);
+        }
+    }
+
+    // All communication between the services should be encrypted using secure communication protocols.
+    private void hasEncrytedInternalConnection(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasDataCharecteristicViolation(node, "Stereotype", "internal", "Stereotype", "encrypted_connection")) {
+            violationsSet.add(8);
+        }
+    }
+
+    // A central logging subsystem which includes a monitoring dashboard should exist.
+    private void hasLoggingServer(AbstractTransposeFlowGraph aTFG, Map<Integer, List<AbstractTransposeFlowGraph>> existenceViolations) {
+        existenceViolations.putIfAbsent(9, new ArrayList<AbstractTransposeFlowGraph>());
+        if (!hasNodeWithCharacteristic(aTFG, "Stereotype", "logging_server")) {
+            existenceViolations.get(9)
+                    .add(aTFG);
+        }
+    }
+
+    // For all microservices, there should exist a local logging agent decoupled from the microservice but deployed
+    // on the same host. Log data from microservices should not be send to the central logging system directly, but
+    // collected by the logging agent, written to a local file, and eventually send to the central system by it.
+    private void hasLocalLogging(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "internal") && !hasNodeCharacteristic(node, "Stereotype", "local_logging")) {
+            violationsSet.add(10);
+            violationsSet.add(11);
+        }
+    }
+
+    // The local logging agent should sanitize the log data and remove any PII, passwords, API keys, etc.
+    private void hasLogSanitization(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "local_logging") && !hasNodeCharacteristic(node, "Stereotype", "log_sanitization")) {
+            violationsSet.add(11);
+        }
+    }
+
+    // A message broker should be used to realize the communication between local logging agent and central logging
+    // system. These two should use mutual authentication and encrypt all transmitted data and availability should
+    // be ensured by providing periodic health and status data.
+    private void hasMessageBroker(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "logging_server")
+                && !hasDataCharacteristicAcrossVariables(node, "Stereotype", "message_broker")) {
+            violationsSet.add(12);
+        }
+    }
+
+    // Secrets should be managed centrally following the Secret as a Service principle.
+    private void hasSecretManager(AbstractTransposeFlowGraph aTFG, Map<Integer, List<AbstractTransposeFlowGraph>> existenceViolations) {
```

Comment:
Replace `aTFG` with `transposeFlowGraph` and `existenceViolations` with the new name mentioned above